src/monodroid/jni/monodroid-glue.c

```diff
@@ -1420,6 +1418,77 @@ java_gc (JNIEnv *env)
 }
 
 static void
+set_bridge_processing_field (MonoVTable *jnienv_vtable, MonoClassField *android_runtime_jnienv_bridge_processing_field, mono_bool value)
+{
+	mono.mono_field_static_set_value (jnienv_vtable, android_runtime_jnienv_bridge_processing_field, &value);
+}
+
+struct MonoDroidDomainId {
+	int domain_id;
+	MonoClass *android_runtime_jnienv;
+	MonoClassField *android_runtime_jnienv_bridge_processing_field;
+	MonoVTable *jnienv_vtable;
+
+	struct MonoDroidDomainId* next;
+};
+
+typedef struct MonoDroidDomainId MonoDroidDomainId;
+MonoDroidDomainId *domains_list;
+
+static void
+add_monodroid_domain (MonoDomain *domain)
+{
+	MonoDroidDomainId *node = calloc (1, sizeof (MonoDroidDomainId));
+
+	/* We need to prefetch all these information prior to using them in gc_cross_reference as all those functions
+	 * use GC API to allocate memory and thus can't be called from within the GC callback as it causes a deadlock
+	 * (the routine allocating the memory waits for the GC round to complete first)
+	 */
+	MonoClass *jnienv = monodroid_get_class_from_name (&mono, domain, "Mono.Android", "Android.Runtime", "JNIEnv");;
+	node->domain_id = mono.mono_domain_get_id (domain);
+	node->android_runtime_jnienv = jnienv;
+	node->android_runtime_jnienv_bridge_processing_field = mono.mono_class_get_field_from_name (jnienv, "BridgeProcessing");
+	node->jnienv_vtable = mono.mono_class_vtable (domain, jnienv);
+	node->next = domains_list;
+
+	domains_list = node;
+}
+
+static void
+remove_monodroid_domain (int domain_id)
+{
+	MonoDroidDomainId *node = domains_list;
+	MonoDroidDomainId *prev = NULL;
+
+	while (node != NULL) {
+		if (node->domain_id != domain_id) {
+			prev = node;
+			node = node->next;
+			continue;
+		}
+
+		if (prev != NULL)
+			prev->next = node->next;
+		else
+			domains_list = node->next;
+
+		free (node);
+
+		break;
+	}
+}
+
+#define FOREACH_DOMAIN(list,statement) do { \
```

Comment:
Given that `FOREACH_DOMAIN` is used precisely *twice*, and that the difference between them is the value to set the field to, I would prefer that this instead be a function:

    static void
    set_bridge_processing (MonoDroidDomainId *list, mono_bool value)
    {
        for ( ; list != NULL; list = list->next ) {
            MonoVTable *vtable = node->stable;
            MonoClassField *field = node->field;
            set_bridge_processing_field (vtable, field, &value);
        }
    }